### PR TITLE
Split bash commands by the new line character

### PR DESCRIPTION
**Pull request by tofarr**
**End-user friendly description of the problem this fixes or functionality that this introduces**

- [X] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below

---
When a bash command spans multiple lines the output is really confusing - so we split each command by the newline character for better readability. My initial approach was to split the command into multiple separate commands and add these to the event stream individually, but this did not play well with the LLM - The behavior was non-deterministic, but sometimes it would basically say "Thank you for running the first command - what about the others?" And it would supply the commands that had not yet been run to be run again, resulting in duplicates in the stream!

So my approach was to improve the output from splitting commands in the runtime client, while still treating these as a single `RunCmdAction`

## Example 1:
```
Print the word "OpenHands" in the bash console in ascii art. Use only the echo command to do this. New lines of text should use the same invocation of the echo command
```
image

## Example 2
```
Please run the following commands all at once:

* ls -lah
* echo "May the force be with you!"
* echo "May the odds be ever in your favor!"
* echo "Live long and prosper!"
```
image

## Example 3
```
Please run the following commands all at once:

* ls -lah
* echo "hello world"
* echo a sample kubernetes deployment yaml into deploy.yaml
```
(This looks a little uglier because the console interprets each tab as 8 spaces.)
image

## Example 4 (With confirmation mode!)
```
Please do the following in a single bash command:

* sleep for 5 seconds
* print "You must Narfle the Garthok!"
* Print the current working directory
```
image

## Example 5 (Running user commands)
image

### CODE REVIEW

**Review by xingyaoww — COMMENTED** [review]
We actually something like this on the backend - what do you feel if we move them to the controller here? That is, we break one CmdRunAction to multiple ones and send them over for execution in agent controller?

https://github.com/All-Hands-AI/OpenHands/blob/ec3152b6e1b96f0f68876cb134d063413f36221c/openhands/runtime/client/client.py#L433-L437

File: openhands/controller/agent_controller.py [hunk 1]
```diff
@@ -449,7 +449,21 @@ async def _step(self) -> None:
...
+                commands = action.command.split('\n')
+                for i, cmd in enumerate(commands):
+                    if not cmd:
+                        continue
+                    new_action = CmdRunAction(command=cmd)
+                    if i < len(commands) - 1:
+                        new_action.thought = ''
+                    else:
+                        new_action.thought = action.thought
+                    self.event_stream.add_event(new_action, EventSource.AGENT)
```
Comment (enyst) [inline openhands/controller/agent_controller.py:464]: This is an interesting idea, how will it work with longer running commands though? Here `_pending_action` is set. When the first new action is executed the runtime will put an obs in the stream which will make it unset [here](https://github.com/All-Hands-AI/OpenHands/pull/4462/files#diff-1e082f415afb40de1d5c2a3fa327326225e1da28a142ad89c9524109f35fb36eR238)... this places them in the stream and continues looping [here](https://github.com/All-Hands-AI/OpenHands/pull/4462/files#diff-1e082f415afb40de1d5c2a3fa327326225e1da28a142ad89c9524109f35fb36eL155). Then the step would pass [here](https://github.com/All-Hands-AI/OpenHands/pull/4462/files#diff-1e082f415afb40de1d5c2a3fa327326225e1da28a142ad89c9524109f35fb36eL394) before all are executed, no? What prevents it?

File: openhands/controller/agent_controller.py [hunk 2]
```diff
@@ -449,7 +449,21 @@ async def _step(self) -> None:
                 == ActionConfirmationStatus.AWAITING_CONFIRMATION
             ):
                 await self.set_agent_state_to(AgentState.AWAITING_USER_CONFIRMATION)
-            self.event_stream.add_event(action, EventSource.AGENT)
+
+            if isinstance(action, CmdRunAction) and '\n' in action.command:
+                # Split the command into multiple CmdRunAction instances
+                commands = action.command.split('\n')
```
Comment (rbren) [inline openhands/controller/agent_controller.py:455]: So we can't just split on `\n`--bash commands are way more complicated than that.

We have some existing logic here to do this better: https://github.com/All-Hands-AI/OpenHands/blob/feee509de7ffe30859611ace9882c72d1a6a3a6b/openhands/runtime/utils/bash.py#L6

**Review by rbren — CHANGES_REQUESTED** [review]

File: openhands/controller/agent_controller.py (hunk 1 above)
Comment (rbren) [reply openhands/controller/agent_controller.py:464]: Yeah this probably messes with _pending_action quite a bit....

@tofarr you can probably test this by asking the agent something like:
> Please run `echo hello` and `ls -lah` in the same `<execute_bash>` block

File: openhands/controller/agent_controller.py (hunk 2 above)
Comment (tofarr) [reply openhands/controller/agent_controller.py:455]: Replaced logic.

**Review by enyst — CHANGES_REQUESTED** [review]
I kept thinking about this PR since the other day, for some odd reason. I'd appreciate if we take some time to make sure this could work. It seems to me it won't work right or I just don't see why it would.

For added fun, let's take a delegates' case: if the action is split in 5 in a delegate controller, and mini-action 1 of 5 has an error, then... okay, I think something else is [bugged](https://github.com/All-Hands-AI/OpenHands/issues/3879) currently, but even with a fix, say like [this](https://github.com/All-Hands-AI/OpenHands/pull/3808/files#diff-1e082f415afb40de1d5c2a3fa327326225e1da28a142ad89c9524109f35fb36eR609), the delegate ends [here](https://github.com/All-Hands-AI/OpenHands/pull/3808/files#diff-1e082f415afb40de1d5c2a3fa327326225e1da28a142ad89c9524109f35fb36eL475). In this PR, is it possible to have obs 2-5 arrive in the stream _after_ the delegate ends? If not, can you please point out what would prevent it?

Because it seems like the child obs could end up in the parent's history, which is a tad messy. 😅

File: openhands/controller/agent_controller.py [hunk 3]
```diff
@@ -449,7 +450,21 @@ async def _step(self) -> None:
...
+
+            if isinstance(action, CmdRunAction):
+                # Split the command into multiple CmdRunAction instances
+                commands = split_bash_commands(action.command)
+                for i, cmd in enumerate(commands):
+                    if not cmd:
+                        continue
+                    new_action = CmdRunAction(command=cmd)
+                    if i < len(commands) - 1:
+                        new_action.thought = ''
```
Comment (rbren) [inline openhands/controller/agent_controller.py:462]: what's this about?

**Comment by enyst** [comment]
It is possible that the action includes several times the same bash command, right? Like `echo *`, I don't know. 

I wonder what would happen [here](https://github.com/All-Hands-AI/OpenHands/blob/91308ba4dc4ab49618f5723269c583c727fe262d/openhands/controller/stuck.py#L82). If they're separate in the stream, like mini-action 1, mini-obs 1, mini-action 2... it seems like they can be detected as repeated actions / obs, and the agent should stop with `stuck in a loop` error. 🤔 We do this for some cases when the LLM gets trapped and responds with the same thing over and over.

**Comment by rbren** [comment]
> It is possible that the action includes several times the same bash command, right? Like echo *, I don't know.

Yeah this is possible, but I can't imagine a scenario where the agent would *want* to do this :)

File: openhands/controller/agent_controller.py (hunk 3 above)
Comment (tofarr) [reply openhands/controller/agent_controller.py:462]: This was from the original prompt:
```.. Each instance should have a blank 'thought' attribute, except for the last one which should have the thought from the original action...```

The reason for this is that it prevents the frontend UI from printing the thought multiple times. e.g.: (Without this functionality):
image

**Comment by rbren** [comment]
So this seems to work on the backend, but the frontend isn't splitting things up correctly

Screenshot 2024-10-21 at 10 51 09 AM

here's my sample prompt:
```
please run the following commands all at once:

* ls -lah
* echo "hello world"
* echo a sample kubernetes deployment yaml into deploy.yaml
```

it seems like we get all the CommandRunActions first, then we get all the Observations. Instead, we should get them collated properly

**Comment by rbren** [comment]
also notice the mangled YAML content on the last command there--not sure what the issue could be...

**Comment by tofarr** [comment]
The command posted by @rbren seems to be working now - though the bash console seems to turn tabs into 8 spaces, which makes it hard to read:
image

**Comment by rbren** [comment]
@tofarr looks like it's still not working right based on your screenshot. The console in the UI should show
```
$ ls -lah
./foo
./bar
$ echo "hello world"
hello world
```
etc. I.e. we should see the output of one command before the start of the next

File: frontend/src/services/actions.ts [hunk 4]
```diff
@@ -56,12 +56,6 @@ const messageActions = {
     if (message.args.thought) {
       store.dispatch(addAssistantMessage(message.args.thought));
     }
-    if (
-      !message.args.is_confirmed ||
-      message.args.is_confirmed !== "rejected"
-    ) {
-      store.dispatch(appendInput(message.args.command));
-    }
```
Comment (tofarr) [inline frontend/src/services/actions.ts:64]: Commands are only outputted on completion

File: openhands/events/observation/commands.py [hunk 5]
```diff
@@ -13,6 +13,7 @@ class CmdOutputObservation(Observation):
     exit_code: int = 0
     hidden: bool = False
     observation: str = ObservationType.RUN
+    meta: str = ''
```
Comment (rbren) [inline openhands/events/observation/commands.py:16]: is there a better name for this?

Comment (enyst) [reply openhands/events/observation/commands.py:16]: Perhaps `location`, or `loc`, or `current_dir`?

Comment (tofarr) [reply openhands/events/observation/commands.py]: I renamed it to `interpreter_details`

**Review by rbren — APPROVED** [review]

**Comment by rbren** [comment]
@enyst we went with a very different implementation here if you wouldn't mind taking another look and unblocking

@xingyaoww would like your approval here as well

**Comment by enyst** [comment]
I agree my concerns were addressed.

**Pull request merged**